packages/shell-api/src/streams.ts

```diff
@@ -0,0 +1,174 @@
+      processStreamProcessor: pipeline,
+      ...(options ? { options } : {}),
+    });
+
+    if (result.ok !== 1) {
+      return result;
+    }
+
+    const { name, cursorId } = result as {
+      name: string;
+      limit: number;
+      cursorId: number;
+    };
+    const sp = this.getProcessor(name);
+
+    const stopAndDrop = () => {
+      sp.stop()
+        .then(() => sp.drop())
+        .catch(() => {
+          /* ignore */
+        });
+    };
+
+    try {
+      await sp._sampleFrom(cursorId);
+    } catch (err) {
+      // try to stop and drop the temp processor on error
+      // wait until execution resumed if its interrupted
+      const isInterrupted = err instanceof MongoshInterruptedError;
+      Promise.resolve(
+        isInterrupted ? this._instanceState.onResumeExecution() : 0
+      )
+        .then(stopAndDrop)
+        .catch(() => void 0);
+
+      throw err;
+    }
+
+    // stop and drop the temp processor if reached the end of sample
+    return stopAndDrop();
+  }
+
+  @returnsPromise
+  async createStreamProcessor(
+    name: string,
+    pipeline: Document[],
+    options?: Document
+  ) {
+    if (typeof name !== 'string' || name.trim() === '') {
+      throw new MongoshInvalidInputError(
+        `Invalid processor name: ${name}`,
+        CommonErrors.InvalidArgument
+      );
+    }
+    if (!Array.isArray(pipeline) || !pipeline.length) {
+      throw new MongoshInvalidInputError(
+        'Invalid pipeline',
+        CommonErrors.InvalidArgument,
+        pipeline
+      );
+    }
+    const result = await this._runStreamCommand({
+      createStreamProcessor: name,
+      pipeline,
+      ...(options ? { options } : {}),
+    });
+
+    if (result.ok !== 1) {
+      return result;
+    }
+
+    return this.getProcessor(name);
+  }
+
+  @returnsPromise
+  async listStreamProcessors(filter: Document) {
+    const result = await this._runStreamCommand({
+      listStreamProcessors: 1,
+      filter,
+    });
+    if (result.ok !== 1) {
+      return result;
+    }
+    const rawProcessors = result.streamProcessors;
+    const sps = rawProcessors.map((sp: StreamProcessor) =>
+      this.getProcessor(sp.name)
+    );
+
+    return Object.defineProperties(sps, {
+      [asPrintable]: { value: () => rawProcessors },
+      [shellApiType]: { value: 'ListStreamProcessorsResult' },
+    });
+  }
+
+  @returnsPromise
+  async listConnections(filter: Document) {
+    return this._runStreamCommand({
+      listStreamConnections: 1,
+      filter,
+    });
+  }
+
+  async _runStreamCommand(cmd: Document, options: Document = {}) {
+    const { _mongo, _instanceState } = this;
+    const interruptable = _instanceState.interrupted.asPromise();
+    try {
+      return await Promise.race([
+        _mongo._serviceProvider.runCommand(ADMIN_DB, cmd, options), // run cmd
+        interruptable.promise, // unless interruppted
+      ]);
+    } finally {
+      interruptable.destroy();
+    }
```

Comment:
@shaketbaby Can you remind me, what was the reason for this logic here, deviating from the core `_runCommand` helper? Is this just to get a specific type of error thrown from this method when execution is interrupted? (Just a question to make sure any changes to the interruption logic to support this here make sense for the streams implementation)

Reply:
I'll be honest that it's been a while and I don't remember all the details. I think this was added mainly for `sample` where we don't really need to destroy the underlying connection and just need to return the control to the end user. But looking again now, I am wondering if this is actually really required for all the other commands other than the sample related ones. Because they don't really catch the exception. For the normal core `_runCommand` helper, if user presses `ctrl + c` while waiting for the server to return the command response, the interrupted error will be thrown and the control will be returned to the user, right? If that's the case, then I think this is not needed for non-sample related comamnds.

Reply:
Basically, I think if there's a way to take over the handling of interruption on `ctrl + c` completely, then that should be good for all the commands we need to support.

Reply:
This is what my suggestion for this would look like, finally got around to it, please take a look if you have the time: https://github.com/mongodb-js/mongosh/pull/1616